chalk-solve/src/wf.rs

```diff
@@ -41,63 +42,63 @@ pub struct WfSolver<'db, I: Interner> {
     solver_choice: SolverChoice,
 }
 
-/// A trait for retrieving all types appearing in some Chalk construction.
-///
-/// FIXME: why is this not a `Folder`?
-trait FoldInputTypes: HasInterner {
-    fn fold(&self, interner: &Self::Interner, accumulator: &mut Vec<Ty<Self::Interner>>);
+struct InputTypeCollector<'i, I: Interner> {
+    types: Vec<Ty<I>>,
+    interner: &'i I,
 }
 
-impl<T: FoldInputTypes> FoldInputTypes for [T] {
-    fn fold(&self, interner: &T::Interner, accumulator: &mut Vec<Ty<T::Interner>>) {
-        for f in self {
-            f.fold(interner, accumulator);
+impl<'i, I: Interner> InputTypeCollector<'i, I> {
+    fn new(interner: &'i I) -> Self {
+        Self {
+            types: Vec::new(),
+            interner,
         }
     }
 }
 
-impl<T: FoldInputTypes> FoldInputTypes for Vec<T> {
-    fn fold(&self, interner: &T::Interner, accumulator: &mut Vec<Ty<T::Interner>>) {
-        for f in self {
-            f.fold(interner, accumulator);
-        }
+impl<'i, 't, I: Interner> Visitor<'i, I> for InputTypeCollector<'i, I> {
+    type Result = ();
+
+    fn as_dyn(&mut self) -> &mut dyn Visitor<'i, I, Result = Self::Result> {
+        self
     }
-}
 
-impl<I: Interner> FoldInputTypes for Parameter<I> {
-    fn fold(&self, interner: &I, accumulator: &mut Vec<Ty<I>>) {
-        if let ParameterKind::Ty(ty) = self.data(interner) {
-            ty.fold(interner, accumulator)
-        }
+    fn interner(&self) -> &'i I {
+        self.interner
     }
-}
 
-impl<I: Interner> FoldInputTypes for Substitution<I> {
-    fn fold(&self, interner: &I, accumulator: &mut Vec<Ty<I>>) {
-        self.parameters(interner).fold(interner, accumulator)
+    fn visit_alias_eq(&mut self, alias_eq: &AliasEq<I>, outer_binder: DebruijnIndex) {
```

Comment:
I guess this is why you added the `AliasEq` callback. Ah, I was imagining before that we would intercept `AliasEq` when we are visiting *goals*, but in this code we are going to visit `WhereClause` values that are outside of any goal or program clause, right? Hmm, that's annoying!

Something about having `visit_alias_eq` doesn't feel right, it's kind of a "random level" in the type hierarchy, and we've kept the "callback points" relatively contained up until now. I guess that having a callback either for `WhereClause` (going "up a level") or for `AliasTy` (going "down a level") feels potentially a bit better to me. An `AliasTy` would also get invoked when walking types, I imagine, which in some sense fits well -- it just adds one more category of type that gets a special callback. (But weird in that it also gets invoked from other contexts...)

That said, I think I'd be ok landing this "as is" with a kind of FIXME to resolve the right callback here -- it's a bit tied into how we are handling implied bounds (that's what this code is doing), and I also think the right answer might become more clear over time.

Reply:
I changed callbacks a little (and added visitor for EnvElaborator), take a look please

Reply:
Will do